### PR TITLE
Crypto: panic if BLS init fails

### PR DESCRIPTION
**Pull request by tarakby**
- BLS initialization is called the very first time any BLS tool is called. 
- The initialization fails if there is source code inconsistency between the C and Go layer (sanity check) or if the C level initialization is not able to allocate memory. 
- If BLS-init fails, there is no way to recover, the function panics. 

Note: a bigger refactor includes separating all BLS tools into a separate package and call the initialization process as part of the `init` function of the package. This is a bigger refactor and will be a longer-term solution.

### CODE REVIEW

**Review by jordanschalm — COMMENTED** [review]
>  a bigger refactor includes separating all BLS tools into a separate package and call the initialization process as part of the init function of the package

I would even do this with the packages as-is. If the `crypto` package as a whole is in a state where calling an innocuous function can cause it panic, it is better panic right away before the program starts.

**Comment by Kay-Zee** [comment]
Yeah, as jordan said, any reason to lazy load? Can just use the built in `init()` function otherwise

**Comment by tarakby** [comment]
@Kay-Zee  @jordanschalm  That's actually part of the large refactor I would like to make, I just added a small part of it to this PR.

File: crypto/bls.go [hunk 1]
```diff
@@ -82,8 +70,8 @@ func (sk *PrKeyBLSBLS12381) Sign(data []byte, kmac hash.Hasher) (Signature, erro
 	// hash the input to 128 bytes
 	h := kmac.ComputeHash(data)
 
-	// intialize BLS context
-	_ = newBLSBLS12381()
+	// set BLS context
+	blsInstance.reInit()
```
Comment (jordanschalm) [inline crypto/bls.go:74]: > If the implementation evolves and relic has multiple contexts,
> reinit should be called at every a. operation.

So this is kind of a safety measure for expected changes in the relic implementation? Or relic configuration?

**Review by jordanschalm — APPROVED** [review]
I'm not clear on why we need `reInit`, but the rest looks good

File: crypto/bls.go (hunk 1 above)
Comment (tarakby) [reply crypto/bls.go:74]: Relic library has settings stored in a context structure. It only supports one context at a time.
If we use Relic for BLS on the specific curve we're using (BLS 12-381), we need to generate and set that context (happens at `init()`). 
We're currently using Relic with only one context, so we could just `init()` once and forget about it. But if we were to add more algorithms or curves, we will have to set the right context before every use. `reInit()` sets that context that is previously generated and stored. `reInit()` is just setting one pointer so it's inexpensive and makes sure there are no problems if we use Relic for other things.

**Pull request merged**